{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1fd1691b-c3fd-428e-aa9f-11e3baddd2f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:19:54.885170Z",
     "iopub.status.busy": "2023-12-26T13:19:54.879842Z",
     "iopub.status.idle": "2023-12-26T13:20:04.245320Z",
     "shell.execute_reply": "2023-12-26T13:20:04.243218Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "## setup libs \n",
    "\n",
    "pkg <- c(\"rtracklayer\", \"tictoc\",\"here\", \"tidyverse\", \"BiocParallel\")\n",
    "\n",
    "invisible(suppressMessages(lapply(c(pkg), library, character.only = TRUE)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2768082-fc2b-4497-b7e6-8b02b3828818",
   "metadata": {},
   "source": [
    "# Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "36319f03-c842-4772-b3f6-c3f63f851b3f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:58:56.438501Z",
     "iopub.status.busy": "2023-12-26T13:58:56.436593Z",
     "iopub.status.idle": "2023-12-26T13:58:58.432912Z",
     "shell.execute_reply": "2023-12-26T13:58:58.430869Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#traindir <- here('code', 'dev', 'data')\n",
    "\n",
    "traindir <- here('data', 'svm')\n",
    "\n",
    "pos <- c('E105H_ATAC', 'E115H_ATAC', 'ENCSR451NAE', 'ENCSR552ABC' ,'ENCSR652CNN','ENCSR068YGC')\n",
    "neg <- c('G4-LIF','ENCSR088UYE', 'ENCSR154BXN', 'ENCSR559FAJ', 'ENCSR302LIV', 'ENCSR551WBK')\n",
    "\n",
    "## load pos & neg pk sets in as lists\n",
    "pos <- lapply(pos, function(file){\n",
    "    path <- file.path(traindir, 'bed', sprintf('%s.mm10.narrowPeak', file))\n",
    "    rtracklayer::import(path,format='narrowPeak')\n",
    "})\n",
    "\n",
    "neg <- lapply(neg, function(file){\n",
    "    path <- file.path(traindir, 'bed',sprintf('%s.mm10.narrowPeak', file))\n",
    "    rtracklayer::import(path,format='narrowPeak')\n",
    "})\n",
    "\n",
    "## load shared invervals generated by bedtools\n",
    "\n",
    "pos_inter <- import.bed(file.path(traindir,'bed','pos_set_multiinter.mm10.bed'))\n",
    "#neg_inter <- import.bed(file.path(traindir,'neg_set_multiinter.mm10.bed'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "678ff899-bbe4-4215-8794-d853d72a0a9a",
   "metadata": {},
   "source": [
    "## Filter peak sets (pos)\n",
    "- filter peak sets (for positive set)\n",
    "    - first filter out unique peaks from each sample\n",
    "    - for multiple overlapping peaks, keeping highest 'score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd80ba6e-9164-41ec-8592-8ecb44d58981",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pos_unique <- do.call('c', lapply(pos, function(gr){\n",
    "    suppressWarnings(gr[!overlapsAny(gr, pos_inter)])\n",
    "}))\n",
    "\n",
    "#length=27848\n",
    "\n",
    "pos_pooled <- suppressWarnings(do.call('c', pos))\n",
    "    pos_pooled <- pos_pooled[overlapsAny(pos_pooled, pos_inter)]\n",
    "ol <- findOverlaps(pos_inter, pos_pooled)\n",
    "\n",
    "#length(unique(queryHits(ol))) #242457\n",
    "\n",
    "#tic()\n",
    "pos_keep <- unique(do.call('c', bplapply(unique(queryHits(ol)), function(i){\n",
    "    ol_idx <- subjectHits(ol)[which(queryHits(ol)==i)]\n",
    "    if(length(ol_idx) > 1){\n",
    "        keep <- ol_idx[which.max(mcols(pos_pooled[ol_idx])$signalValue)]\n",
    "        return(keep)\n",
    "    }\n",
    "},  BPPARAM=MulticoreParam(workers = 50))))\n",
    "\n",
    "#toc() 217.578s\n",
    "\n",
    "pos_final <- c(pos_unique, pos_pooled[pos_keep]) \n",
    "\n",
    "#length(pos_final) #87239"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af4949ce-6086-4a52-bf27-629f8da4e232",
   "metadata": {},
   "source": [
    "- filtering for neg set not necessary ⬇️"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9efe0bfc-b40c-4739-b695-a62664547ac7",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# neg_unique <- do.call('c', lapply(neg, function(gr){\n",
    "#     suppressWarnings(gr[!overlapsAny(gr, neg_inter)])\n",
    "# }))\n",
    "\n",
    "# neg_pooled <- suppressWarnings(do.call('c', neg))\n",
    "#     neg_pooled <- neg_pooled[overlapsAny(neg_pooled, neg_inter)]\n",
    "# ol2 <- findOverlaps(neg_inter, neg_pooled)\n",
    "\n",
    "# ## for multiple overlap, get peak with highest signal score\n",
    "# #tic()\n",
    "# neg_keep <- unique(do.call('c', bplapply(unique(queryHits(ol)), function(i){\n",
    "#     ol_idx <- subjectHits(ol2)[which(queryHits(ol2)==i)]\n",
    "#     if(length(ol_idx) > 1){\n",
    "#         keep <- ol_idx[which.max(mcols(neg_pooled[ol_idx])$signalValue)]\n",
    "#         return(keep)\n",
    "#     }\n",
    "# },  BPPARAM=MulticoreParam(workers = 50))))\n",
    "\n",
    "# #toc() #202.9\n",
    "\n",
    "# neg_final <- c(neg_unique, neg_pooled[neg_keep])\n",
    "\n",
    "# #length(neg_final) #154187\n",
    "\n",
    "# ## trim neg set to 500bp by summit\n",
    "# start(neg_final) <- start(neg_final) + neg_final$peak - 250\n",
    "# neg_final <- resize(neg_final, 500, 'start')\n",
    "\n",
    "# ## neg set = 2kb away from any pos peak\n",
    "# neg_final <- neg_final[!overlapsAny(neg_final, \n",
    "#                                     resize(pos_final, 2000, 'center'))]\n",
    "\n",
    "\n",
    "# length(neg_final) #111852\n",
    "# ## subsample to get same number of peak between 2 sets\n",
    "# #set.seed <- 611\n",
    "# #neg_final.ss <- sample(neg_final, length(pos_final))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3780c7f-3a38-4ee6-9f56-b146dfa50bcc",
   "metadata": {},
   "source": [
    "- filter pos set to be outside 2kb of annotated TSS (from EPD3, mm10), i.e. prioritize enhancer-like signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85412b8c-c233-41f7-855d-eac8a8a25af8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prom <- resize(import.bed(here('_data', 'references', 'mm10_epd3.bed')), width = 1000, fix = 'center')\n",
    "\n",
    "## trim pos_final to 500bp around summit\n",
    "start(pos_final) <- start(pos_final) + pos_final$peak - 250\n",
    "pos_final <- resize(pos_final, 500, 'start')\n",
    "\n",
    "pos_final <- pos_final[!overlapsAny(pos_final, prom, maxgap = 2000)]\n",
    "\n",
    "#length(pos_final) #64136"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2518e05c-7634-4700-afba-c1a89d89f8b7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:17:09.156477Z",
     "iopub.status.busy": "2023-12-26T13:17:09.154620Z",
     "iopub.status.idle": "2023-12-26T13:17:09.603838Z",
     "shell.execute_reply": "2023-12-26T13:17:09.601746Z"
    }
   },
   "outputs": [],
   "source": [
    "export.bed(pos_final, \n",
    "          file.path(traindir,'bed', 'pos_set_filtered_500bp.mm10.bed'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34d9844-142c-48b7-a031-b154a131017d",
   "metadata": {},
   "source": [
    "## Generate null seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d914cc99-3a31-49d8-9bbd-dcead36ae336",
   "metadata": {},
   "source": [
    "- generate matching GC & repeats null seqs from the genome & get those overlapping with neg peaks as final negset\n",
    "- adapt genNullseq function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e585d472-e76f-4cc5-9ea9-9b4c145c0592",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:29:49.393768Z",
     "iopub.status.busy": "2023-12-26T13:29:49.392203Z",
     "iopub.status.idle": "2023-12-26T13:29:49.408210Z",
     "shell.execute_reply": "2023-12-26T13:29:49.406419Z"
    },
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "## adapted from https://rdrr.io/cran/gkmSVM/src/R/genNullSeqs.R\n",
    "\n",
    "genNullSeqs = function(\n",
    "  inputBed, \n",
    "  xfold = 1, \n",
    "  repeat_match_tol = 0.02, \n",
    "  GC_match_tol = 0.02, \n",
    "  length_match_tol = 0.02, \n",
    "  batchsize = 5000, \n",
    "  nMaxTrials = 20, \n",
    "  genome = NULL   \n",
    "){\n",
    "  \n",
    "  \n",
    "  if (requireNamespace(\"GenomicRanges\", quietly = TRUE)&\n",
    "      requireNamespace(\"rtracklayer\", quietly = TRUE)&\n",
    "      requireNamespace(\"BiocGenerics\", quietly = TRUE)&\n",
    "      requireNamespace(\"Biostrings\", quietly = TRUE)&\n",
    "      requireNamespace(\"GenomeInfoDb\", quietly = TRUE)&\n",
    "      requireNamespace(\"IRanges\", quietly = TRUE)&\n",
    "      requireNamespace(\"S4Vectors\", quietly = TRUE)){\n",
    "    \n",
    "    \n",
    "    #inputBedFile = '~/Downloads/ctcfpos.bed'\n",
    "    #xfold = 1; \n",
    "    # if(is.null(genome)){  \n",
    "    #   if(toupper(genomeVersion)=='HG18'){\n",
    "    #     if(requireNamespace(\"BSgenome.Hsapiens.UCSC.hg18.masked\", quietly = TRUE)){\n",
    "    #       genome <- BSgenome.Hsapiens.UCSC.hg18.masked::BSgenome.Hsapiens.UCSC.hg18.masked\n",
    "    #     }\n",
    "    #   } else{\n",
    "    #     if(requireNamespace(\"BSgenome.Hsapiens.UCSC.hg19.masked\", quietly = TRUE)){\n",
    "    #       genome <- BSgenome.Hsapiens.UCSC.hg19.masked::BSgenome.Hsapiens.UCSC.hg19.masked\n",
    "    #     }\n",
    "    #   }\n",
    "    # } else {\n",
    "        #genome <- BSgenome.Sscrofa.UCSC.susScr3.masked::BSgenome.Sscrofa.UCSC.susScr3.masked\n",
    "    # }\n",
    "    \n",
    "    seqnams= GenomeInfoDb::seqnames(genome)\n",
    "    chrlens = GenomeInfoDb::seqlengths(genome)\n",
    "    chrpos = cumsum(as.numeric(chrlens))\n",
    "    pmax = max(chrpos)\n",
    "    chrpos = c(chrpos,1E12)\n",
    "    chrpos0 = c(0, chrpos)\n",
    "    ichrA = as.character(names(chrlens)); \n",
    "    \n",
    "    \n",
    "    getichrpos = function(ipos){\n",
    "      j = order(ipos); \n",
    "      ipos = sort(ipos); \n",
    "      ci = 1;\n",
    "      res = rep(NA, length(ipos))\n",
    "      for(i in 1:length(ipos))\n",
    "      {\n",
    "        while(ipos[i]>chrpos[ci]){\n",
    "          ci = ci+1; \n",
    "        }\n",
    "        res[j[i]] = ci\n",
    "      }\n",
    "      return(res); \n",
    "    }\n",
    "    \n",
    "    generateRandomGenSeqs = function(seqlens){\n",
    "      rpos = sample(pmax, length(seqlens), replace = TRUE)\n",
    "      ichr1 = getichrpos(rpos)\n",
    "      ichr2 = getichrpos(rpos+seqlens)\n",
    "      \n",
    "      jj = which(ichr1!=ichr2)\n",
    "      while(length(jj)>0){\n",
    "        rpos[jj] = sample(pmax, length(jj), replace = TRUE)\n",
    "        ichr1 = getichrpos(rpos)\n",
    "        ichr2 = getichrpos(rpos+seqlens)\n",
    "        jj = which(ichr1!=ichr2)\n",
    "      }\n",
    "      chr = ichrA[ichr1]\n",
    "      start = rpos - chrpos0[ichr1];\n",
    "      names <- chr; \n",
    "      ranges <- IRanges::IRanges(start=start, width=seqlens)\n",
    "      strand <- BiocGenerics::strand(sample(c(\"+\", \"-\"), length(names), replace=TRUE))\n",
    "      gr <- GenomicRanges::GRanges(seqnames=names, ranges=ranges, strand=strand)\n",
    "    }\n",
    "    \n",
    "    #inBed = rtracklayer::import.bed(inputBedFN)\n",
    "     inBed = inputBed \n",
    "    \n",
    "    #check the BED file:\n",
    "    inbed = GenomicRanges::as.data.frame(inBed)\n",
    "    jj = which(is.na(match(as.character(inbed$seqnames),as.character(seqnams))))\n",
    "    if (length(jj)>0){\n",
    "      cat( paste('ERROR: Chromosome name not recognized for',length(jj), 'sequences.\\n'))\n",
    "      cat( unique(as.character(inbed$seqnames[jj])))\n",
    "      return(NULL)\n",
    "    }\n",
    "    jj = which(inbed$end>GenomeInfoDb::seqlengths(genome)[as.character(inbed$seqnames)])\n",
    "    if (length(jj)>0){\n",
    "      cat( 'ERROR: Region outside chromosome. (Check the genome version) \\n')\n",
    "      print( inbed[jj,])\n",
    "      return(NULL)\n",
    "    }\n",
    "    \n",
    "    gcContent <- function(seqs)\n",
    "    {\n",
    "      alf <- Biostrings::alphabetFrequency(seqs, as.prob=TRUE)\n",
    "      gc = rowSums(alf[,c(\"G\", \"C\"), drop=FALSE])\n",
    "    }\n",
    "    # \n",
    "    # gc1=desGC[unmatched]\n",
    "    # gc2=rndGC\n",
    "    # len1=desLens[unmatched]\n",
    "    # len2=width(rndBed)\n",
    "    # rpt1=desRpt[unmatched]\n",
    "    # rpt2=rndRpt\n",
    "    # gc_th = GC_match_tol,\n",
    "    # len_th = repeat_match_tol,\n",
    "    # rpt_th = length_match_tol)\n",
    "    \n",
    "    matchSeqs = function(gc1, gc2, len1, len2, rpt1, rpt2,  gc_th=0.02, len_th=0.02, rpt_th=0.02){\n",
    "      ##\n",
    "      #  gc1 = rndGC; \n",
    "      #  gc2 = inGC; \n",
    "      #  len1 = seqlens; \n",
    "      #  len2 = seqlens;\n",
    "      #  rpt1 = inRpt; \n",
    "      #  rpt2 = rndRpt; \n",
    "      #  gc1=desGC[unmatched];gc2= rndGC;len1= desLens[unmatched];len2= BiocGenerics::width(rndBed);rpt1= desRpt[unmatched];rpt2= rndRpt;\n",
    "                      \n",
    "      #  gc_th=0.02\n",
    "      #  len_th=0.02\n",
    "      \n",
    "      \n",
    "      # gc1, len1 are the desired \n",
    "      # gc2 and len2 are to be matched \n",
    "      len_th = len_th * len1; \n",
    "      \n",
    "      i1 = order(gc1)\n",
    "      i2 = order(gc2)\n",
    "      \n",
    "      gc1 = gc1[i1]\n",
    "      gc2 = gc2[i2]\n",
    "      len1 = len1[i1]\n",
    "      len2 = len2[i2]\n",
    "      rpt1 = rpt1[i1]\n",
    "      rpt2 = rpt2[i2]\n",
    "      \n",
    "      gc2 = c(gc2, 1E10)\n",
    "      \n",
    "      len_th = len_th[i1]\n",
    "      \n",
    "      m2 = 1; \n",
    "      N = length(i1); \n",
    "      N2 = length(i2);\n",
    "      mtc1 = rep(NA, N)\n",
    "      mtc2 = rep(0, length(i2))\n",
    "      for(i in 1:N){\n",
    "        #if(i%%1000==0){cat(i,' ')}\n",
    "        gc1i = gc1[i]; \n",
    "        len1i = len1[i]\n",
    "        rpt1i = rpt1[i]\n",
    "        len_thi = len_th[i]\n",
    "        \n",
    "        while(gc1i - gc2[m2]>gc_th) {\n",
    "          m2 = m2+1; \n",
    "        }\n",
    "        if(m2<=N2){\n",
    "          m2b=m2;\n",
    "          while(gc2[m2b]-gc1i<=gc_th){\n",
    "            if ((mtc2[m2b]==0)&(abs(len1i-len2[m2b])<=len_thi)&(abs(rpt1i-rpt2[m2b])<=rpt_th)){\n",
    "              mtc2[m2b]=i; \n",
    "              mtc1[i]=m2b;\n",
    "              if(m2b==m2){m2 = m2+1;}\n",
    "              break; \n",
    "            }\n",
    "            m2b =m2b+1; \n",
    "          }\n",
    "        }else{break;}\n",
    "      }\n",
    "      \n",
    "      mtc1 = i2[mtc1]\n",
    "      res = rep(NA, N)\n",
    "      res[i1] = mtc1; \n",
    "      return(res)\n",
    "    }\n",
    "    \n",
    "    #bed = rndBed\n",
    "    repeatRat = function(bed){\n",
    "      chrs = unique(GenomeInfoDb::seqnames(bed))\n",
    "      rpts = rep(0, length(bed))\n",
    "      \n",
    "      for(ichr in as.character(chrs)){\n",
    "        seq = genome[[ichr]]\n",
    "        rpt = Biostrings::masks(seq)[['TRF']]\n",
    "        \n",
    "        jj = which(as.character(GenomeInfoDb::seqnames(bed))==ichr)\n",
    "#       jj = which((GenomeInfoDb::seqnames(bed))==ichr)\n",
    "        \n",
    "        if (length(jj)>0){\n",
    "          \n",
    "          jbed = bed[jj]\n",
    "          jrpts = rep(0, length(jj))\n",
    "          \n",
    "          olaps <- IRanges::findOverlaps(rpt, jbed@ranges)\n",
    "          #isect <- pintersect(rpt[queryHits(olaps)], jbed@ranges[subjectHits(olaps)])\n",
    "          \n",
    "          qdf= GenomicRanges::as.data.frame(rpt)[S4Vectors::queryHits(olaps),]\n",
    "          isect <- IRanges::pintersect(IRanges::IRanges(start=qdf$start,end=qdf$end), jbed@ranges[S4Vectors::subjectHits(olaps)])\n",
    "          \n",
    "          jres = S4Vectors::subjectHits(olaps)\n",
    "          olap_width=BiocGenerics::width(isect)\n",
    "          \n",
    "          ## this could be done faster if duplicated(jres) is empty \n",
    "          for(i in 1:length(jres)){\n",
    "            jrpts[jres[i]]=jrpts[jres[i]]+olap_width[i]\n",
    "          } \n",
    "          rpts[jj] = jrpts;\n",
    "        }\n",
    "      }\n",
    "      \n",
    "      rpts = rpts/BiocGenerics::width(bed)\n",
    "    }\n",
    "    \n",
    "    #check the BED file:\n",
    "    inbed = GenomicRanges::as.data.frame(inBed)\n",
    "    \n",
    "    cat(' importing sequences from', GenomeInfoDb::bsgenomeName(genome),'\\n')\n",
    "    #extract sequences\n",
    "    inSeqs = Biostrings::getSeq(genome, inBed)\n",
    "    seqlens = inbed$width\n",
    "    inGC = gcContent(inSeqs)\n",
    "    cat(' calculating repeat distributions\\n')\n",
    "    inRpt = repeatRat(inBed)\n",
    "    \n",
    "    nout = round(nrow(inbed)*xfold)\n",
    "    #outbed=as.data.frame(matrix(ncol=ncol(inbed), nrow=nout))\n",
    "    outbed=matrix(ncol=ncol(inbed), nrow=nout)\n",
    "    outSeq = rep(inSeqs, length=nout); \n",
    "    \n",
    "    colnames(outbed)=colnames(inbed)\n",
    "    \n",
    "    unmatched = 1:length(outSeq)\n",
    "    \n",
    "    desGC = rep(inGC, length=nout); #desired output GC\n",
    "    desRpt = rep(inRpt, length=nout); #desired output repeat\n",
    "    desLens = rep(seqlens, length=nout); #desired output lengths \n",
    "    \n",
    "    for(iter in 1:nMaxTrials){\n",
    "      if(length(unmatched)>0){\n",
    "        cat(' Trial',iter,'out of',nMaxTrials,'\\n')\n",
    "        rndBed = generateRandomGenSeqs(rep(desLens[unmatched],length.out=batchsize))\n",
    "        rndbed= GenomicRanges::as.data.frame(rndBed)\n",
    "        cat(' importing sequences\\n')\n",
    "        rndSeqs = Biostrings::getSeq(genome, rndBed)\n",
    "        rndGC = gcContent(rndSeqs)\n",
    "        cat(' calculating repeat distributions\\n')\n",
    "        rndRpt = repeatRat(rndBed)\n",
    "        cat(' matching sequences\\n')\n",
    "        \n",
    "        mtc = matchSeqs(desGC[unmatched], rndGC, desLens[unmatched], BiocGenerics::width(rndBed), desRpt[unmatched], rndRpt,\n",
    "                        gc_th = GC_match_tol,\n",
    "                        len_th = length_match_tol,\n",
    "                        rpt_th = repeat_match_tol)\n",
    "        jj = which(!is.na(mtc))\n",
    "        if(length(jj)>0){\n",
    "          #outbed[unmatched[jj],]=rndbed[mtc[jj],];\n",
    "          outbed[unmatched[jj],1:5]=as.matrix(rndbed[mtc[jj],]);\n",
    "          outSeq[unmatched[jj],]=rndSeqs[mtc[jj],];\n",
    "          unmatched = unmatched[-jj]\n",
    "        }\n",
    "        cat(nrow(outbed) - length(unmatched),'sequences found so far, ',length(unmatched), ' remaining.\\n')\n",
    "      }\n",
    "    }  \n",
    "    \n",
    "    if(length(unmatched)>0){\n",
    "      outbed = outbed[-unmatched,]\n",
    "      outSeq = outSeq[-unmatched,]\n",
    "    }\n",
    "\n",
    "    outbed = gsub(' ','', outbed)\n",
    "    #utils::write.table(as.matrix(outbed[,1:3]),quote = FALSE, sep='\\t',row.names = FALSE, col.names = FALSE , file = outputBedFN)  \n",
    "    # if(requireNamespace(\"seqinr\", quietly = TRUE)){\n",
    "    #   outseqnams = paste(outbed[,1], outbed[,2], outbed[,3], 'neg', 1:nrow(outbed), sep='_')\n",
    "    #   seqinr::write.fasta(sequences = sapply(as.character(outSeq), strsplit,''), names =outseqnams,file.out =   outputNegFastaFN); \n",
    "    #   inseqnams = paste(as.character(inbed[,1]), inbed[,2], inbed[,3], 'pos', 1:nrow(inbed), sep='_')\n",
    "    #   seqinr::write.fasta(sequences = sapply(as.character(inSeqs), strsplit,''), names =inseqnams, file.out =   outputPosFastaFN); \n",
    "    # }\n",
    "    return(outbed)\n",
    "    \n",
    "      }else{\n",
    "        cat('\\n \"GenomicRanges\" and \"rtracklayer\" packages are needed. \\n')\n",
    "      }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3abf1a9-48a9-4093-8b8b-2cdfd2718a71",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Mmusculus.UCSC.mm10.TRF.masked)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "06b232c4-a5af-49fd-8caa-118833e4ba95",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:47:01.071091Z",
     "iopub.status.busy": "2023-12-26T13:47:01.067718Z",
     "iopub.status.idle": "2023-12-26T13:47:01.105598Z",
     "shell.execute_reply": "2023-12-26T13:47:01.101586Z"
    }
   },
   "outputs": [],
   "source": [
    "#tic()\n",
    "matchingNull_bed <- bplapply(split(pos_final, seqnames(pos_final)), function(chrom){\n",
    "    genNullSeqs(inputBed = chrom, \n",
    "                genome = BSgenome.Mmusculus.UCSC.mm10.TRF.masked::BSgenome.Mmusculus.UCSC.mm10.TRF.masked, \n",
    "                xfold = 10)\n",
    "}, BPPARAM=MulticoreParam(workers = 43))\n",
    "#toc() #564.433 sec elapsed\n",
    "\n",
    "\n",
    "\n",
    "matchingNull_bed <- data.frame(do.call('rbind', unname(matchingNull_bed)))\n",
    "\n",
    "#nrow(matchingNull_bed ) #481328"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "998c6c40-e8fa-440d-900b-1c68f3693ea7",
   "metadata": {},
   "source": [
    "- convert to GRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "db421108-6475-4f42-a35d-67e011965cec",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:57:21.376445Z",
     "iopub.status.busy": "2023-12-26T13:57:21.373748Z",
     "iopub.status.idle": "2023-12-26T13:57:22.121509Z",
     "shell.execute_reply": "2023-12-26T13:57:22.119181Z"
    }
   },
   "outputs": [],
   "source": [
    "matchingNull_gr <- GRanges(seqnames = matchingNull_bed$seqnames,\n",
    "                           ranges = IRanges(start = as.numeric(matchingNull_bed$start), width = 500), \n",
    "                           strand = rep('*', nrow(matchingNull_bed)), \n",
    "                           name = paste0('RND_', seq_along(matchingNull_bed$name)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "65efb85c-fff3-4fb4-9f0a-98857ebc7697",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T13:57:26.262236Z",
     "iopub.status.busy": "2023-12-26T13:57:26.260509Z",
     "iopub.status.idle": "2023-12-26T13:57:26.285597Z",
     "shell.execute_reply": "2023-12-26T13:57:26.283773Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 1 metadata column:\n",
       "      seqnames              ranges strand |        name\n",
       "         <Rle>           <IRanges>  <Rle> | <character>\n",
       "  [1]    chr10   93821860-93822359      * |       RND_1\n",
       "  [2]     chr2 138144687-138145186      * |       RND_2\n",
       "  [3]    chr14   32990543-32991042      * |       RND_3\n",
       "  [4]    chr15   75736095-75736594      * |       RND_4\n",
       "  [5]    chr17   42677890-42678389      * |       RND_5\n",
       "  [6]     chr4   22504669-22505168      * |       RND_6\n",
       "  -------\n",
       "  seqinfo: 236 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(matchingNull_gr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08378c50-18bf-410e-bc6f-0820043bad92",
   "metadata": {},
   "source": [
    "- After filtering out null-seq within 2kb of pos peaks, options are:\n",
    "    - (1) get those overlapping with neg ATAC peaks from limb, brain, G4\n",
    "    - (2) get those overlapping with ENCODE cCREs\n",
    "    - (3) because ENCODE cCRES did not include embryonic data, combine (1) & (2) !![THIS]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "723dfae6-937e-4159-a28d-b1a434976945",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T14:24:42.995733Z",
     "iopub.status.busy": "2023-12-26T14:24:42.994265Z",
     "iopub.status.idle": "2023-12-26T14:24:43.077934Z",
     "shell.execute_reply": "2023-12-26T14:24:43.076133Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ccres <- import.bed(here('_data', 'references', 'ENCODE_cCRE.mm10.bed'))\n",
    "#length(ccres) #368121\n",
    "neg_ccre <- c(neg, ccres)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec78c9ed-0366-4b3f-a76f-d006635954b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "null_final <- suppressWarnings(matchingNull_gr[!overlapsAny(matchingNull_gr, pos_final, maxgap = 2000)])\n",
    "#length(null_final) #403041\n",
    "null_final <- suppressWarnings(null_final[overlapsAny(null_final, neg_ccre, minoverlap = 0.1*500)])\n",
    "#length(null_final) #67218"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb88dfd3-55f1-4a57-b15a-b39bb4a227f7",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "fd312f56-416b-43e8-a0a6-1b3c44189949",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-26T14:30:13.899582Z",
     "iopub.status.busy": "2023-12-26T14:30:13.898185Z",
     "iopub.status.idle": "2023-12-26T14:30:14.872460Z",
     "shell.execute_reply": "2023-12-26T14:30:14.870685Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "export.bed(pos_final, \n",
    "          file.path(traindir,'bed', 'pos_set_filtered_500bp.mm10.bed'))\n",
    "\n",
    "export.bed(null_final, \n",
    "          file.path(traindir,'bed','neg_set_GC-rpt-matched_500bp.mm10.bed'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d5ac59fc-726e-4009-b40b-07f434ae794c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:13:40.013761Z",
     "iopub.status.busy": "2023-12-25T16:13:40.010457Z",
     "iopub.status.idle": "2023-12-25T16:13:40.054682Z",
     "shell.execute_reply": "2023-12-25T16:13:40.050587Z"
    }
   },
   "outputs": [],
   "source": [
    "set.seed <- 666\n",
    "test.neg <- sample(neg_pooled, 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f11dc57-0483-48f3-b575-7f7523339906",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:15:13.307336Z",
     "iopub.status.busy": "2023-12-25T16:15:13.304022Z",
     "iopub.status.idle": "2023-12-25T16:15:13.330167Z",
     "shell.execute_reply": "2023-12-25T16:15:13.326380Z"
    }
   },
   "source": [
    "- instead of chrom/seqlengths as max length, take width of each pk as max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05569d7a-74e1-417a-9622-a2be6f8e601f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "\n",
    "mm10 <- BSgenome.Mmusculus.UCSC.mm10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "ee7073a8-5801-4050-885e-71e154b525ff",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:52:25.604411Z",
     "iopub.status.busy": "2023-12-25T16:52:25.601084Z",
     "iopub.status.idle": "2023-12-25T16:52:25.634315Z",
     "shell.execute_reply": "2023-12-25T16:52:25.630364Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "getichrpos = function(ipos){\n",
    "  j = order(ipos); \n",
    "  ipos = sort(ipos); \n",
    "  ci = 1;\n",
    "  res = rep(NA, length(ipos))\n",
    "  for(i in 1:length(ipos))\n",
    "  {\n",
    "    while(ipos[i]>chrpos[ci]){\n",
    "      ci = ci+1; \n",
    "    }\n",
    "    res[j[i]] = ci\n",
    "  }\n",
    "  return(res); \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2157ee6b-12c1-4443-84d9-ad0ce4d1c566",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:52:39.822686Z",
     "iopub.status.busy": "2023-12-25T16:52:39.819379Z",
     "iopub.status.idle": "2023-12-25T16:52:39.845533Z",
     "shell.execute_reply": "2023-12-25T16:52:39.841629Z"
    }
   },
   "outputs": [],
   "source": [
    "rpos <- sample(pmax, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f43ac709-62c9-4c2a-86ca-e272a547ea67",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:53:28.959240Z",
     "iopub.status.busy": "2023-12-25T16:53:28.956082Z",
     "iopub.status.idle": "2023-12-25T16:53:29.017700Z",
     "shell.execute_reply": "2023-12-25T16:53:29.013993Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>16</li><li>5</li><li>12</li><li>5</li><li>14</li><li>14</li><li>7</li><li>13</li><li>8</li><li>13</li><li>20</li><li>5</li><li>7</li><li>18</li><li>8</li><li>3</li><li>15</li><li>7</li><li>6</li><li>17</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 16\n",
       "\\item 5\n",
       "\\item 12\n",
       "\\item 5\n",
       "\\item 14\n",
       "\\item 14\n",
       "\\item 7\n",
       "\\item 13\n",
       "\\item 8\n",
       "\\item 13\n",
       "\\item 20\n",
       "\\item 5\n",
       "\\item 7\n",
       "\\item 18\n",
       "\\item 8\n",
       "\\item 3\n",
       "\\item 15\n",
       "\\item 7\n",
       "\\item 6\n",
       "\\item 17\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 16\n",
       "2. 5\n",
       "3. 12\n",
       "4. 5\n",
       "5. 14\n",
       "6. 14\n",
       "7. 7\n",
       "8. 13\n",
       "9. 8\n",
       "10. 13\n",
       "11. 20\n",
       "12. 5\n",
       "13. 7\n",
       "14. 18\n",
       "15. 8\n",
       "16. 3\n",
       "17. 15\n",
       "18. 7\n",
       "19. 6\n",
       "20. 17\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 16  5 12  5 14 14  7 13  8 13 20  5  7 18  8  3 15  7  6 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getichrpos(rpos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e5556c20-c8fe-41d0-b1ad-8c511ea03cc3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:25:43.116459Z",
     "iopub.status.busy": "2023-12-25T16:25:43.113145Z",
     "iopub.status.idle": "2023-12-25T16:25:43.147013Z",
     "shell.execute_reply": "2023-12-25T16:25:43.143297Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1110742384</li><li>257091909</li><li>714336825</li><li>1408754075</li><li>2650870912</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1110742384\n",
       "\\item 257091909\n",
       "\\item 714336825\n",
       "\\item 1408754075\n",
       "\\item 2650870912\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1110742384\n",
       "2. 257091909\n",
       "3. 714336825\n",
       "4. 1408754075\n",
       "5. 2650870912\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1110742384  257091909  714336825 1408754075 2650870912"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample(pmax, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "57246555-7f80-4c2b-957f-0e9ed709e901",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-12-25T16:17:44.614059Z",
     "iopub.status.busy": "2023-12-25T16:17:44.612424Z",
     "iopub.status.idle": "2023-12-25T16:17:44.656770Z",
     "shell.execute_reply": "2023-12-25T16:17:44.653048Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "971"
      ],
      "text/latex": [
       "971"
      ],
      "text/markdown": [
       "971"
      ],
      "text/plain": [
       "[1] 971"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cumsum(width(test.neg[1]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R-4.2.1",
   "language": "R",
   "name": "ir42"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
